readthedocs/projects/validators.py

```diff
@@ -98,3 +99,69 @@ class SubmoduleURLValidator(RepositoryURLValidator):
 
 validate_repository_url = RepositoryURLValidator()
 validate_submodule_url = SubmoduleURLValidator()
+
+
+def validate_build_config_file(path):
+    """
+    Validate that user input is a good relative repository path.
+
+    By 'good', we mean that it's a valid unix path,
+    but not all valid unix paths are good repository paths.
+
+    This validator checks for common mistakes.
+
+    :param valid_filenames: A list of valid file names or None.
+    """
+    invalid_characters = "[]{}()`'\"\\%&<>|,"
+    valid_filenames = [".readthedocs.yaml"]
+    if path.startswith("/"):
+        raise ValidationError(
+            _(
+                "Use a relative path. It should not begin with '/'. "
+                "The path is relative to the root of your repository."
+            ),
+            code="path_invalid",
+        )
+    if path.endswith("/"):
+        raise ValidationError(
+            _("The path cannot end with '/', as it cannot be a directory."),
+            code="path_invalid",
+        )
+    if ".." in path:
+        raise ValidationError(
+            _("Found invalid sequence in path: '..'"),
+            code="path_invalid",
+        )
+    if any(ch in path for ch in invalid_characters):
+        raise ValidationError(
+            mark_safe(
+                _(
+                    "Found invalid character. Avoid these characters: "
+                    "<code>{invalid_characters}</code>"
+                ).format(invalid_characters=invalid_characters),
+            ),
+            code="path_invalid",
+        )
+    # If we should validate filenames
+    if valid_filenames:
```

Comment:
This `if` seems unneeded since it's always statically defined at the beginning of the function. We should remove it and un-indent the code.

Reply:
Ah yes, `valid_filenames` used to be an argument that would make the validator reusable for many purposes, including conf.py (the `conf_py_file` field). But I scrapped that.

Removing :+1: